services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/delete/actions/RemoveRulesAction.java

```diff
@@ -0,0 +1,157 @@
+import org.openkilda.wfm.topology.flowhs.fsm.delete.FlowDeleteFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.delete.FlowDeleteFsm.State;
+import org.openkilda.wfm.topology.flowhs.service.FlowCommandBuilder;
+import org.openkilda.wfm.topology.flowhs.service.FlowCommandBuilderFactory;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class RemoveRulesAction extends FlowProcessingAction<FlowDeleteFsm, State, Event, FlowDeleteContext> {
+    private final FlowResourcesManager resourcesManager;
+    private final FlowCommandBuilderFactory commandBuilderFactory;
+
+    public RemoveRulesAction(PersistenceManager persistenceManager, FlowResourcesManager resourcesManager) {
+        super(persistenceManager);
+        this.resourcesManager = resourcesManager;
+        commandBuilderFactory = new FlowCommandBuilderFactory(resourcesManager);
+    }
+
+    @Override
+    protected void perform(State from, State to, Event event, FlowDeleteContext context, FlowDeleteFsm stateMachine) {
+        Flow flow = getFlow(stateMachine.getFlowId());
+        Set<PathId> pathIds = new HashSet<>(flow.getFlowPathIds());
+
+        FlowCommandBuilder commandBuilder = commandBuilderFactory.getBuilder(flow.getEncapsulationType());
+        Collection<RemoveRule> commands = new ArrayList<>();
+
+        FlowPath forward = flow.getForwardPath();
+        FlowPath reverse = flow.getReversePath();
+        if (forward != null && reverse != null) {
+            pathIds.remove(forward.getPathId());
+            pathIds.remove(reverse.getPathId());
+
+            stateMachine.addFlowResources(buildResources(flow, forward, reverse));
+
+            commands.addAll(commandBuilder.createRemoveNonIngressRules(
+                    stateMachine.getCommandContext(), flow, forward, reverse));
+            commands.addAll(commandBuilder.createRemoveIngressRules(
+                    stateMachine.getCommandContext(), flow, forward, reverse));
+        }
+
+        FlowPath protectedForward = flow.getProtectedForwardPath();
+        FlowPath protectedReverse = flow.getProtectedReversePath();
+        if (protectedForward != null && protectedReverse != null) {
+            pathIds.remove(protectedForward.getPathId());
+            pathIds.remove(protectedReverse.getPathId());
+
+            stateMachine.addFlowResources(buildResources(flow, protectedForward, protectedReverse));
+
+            commands.addAll(commandBuilder.createRemoveNonIngressRules(
+                    stateMachine.getCommandContext(), flow, protectedForward, protectedReverse));
+            commands.addAll(commandBuilder.createRemoveIngressRules(
+                    stateMachine.getCommandContext(), flow, protectedForward, protectedReverse));
```

Comment:
AFAIK protected paths have on Ingress rules. Only transit and egress

Reply:
Right. Fixed